api/rest/cluster_bundle_handler.go

```diff
@@ -0,0 +1,360 @@
+}
+
+// Delete will delete a given bundle, proxying the call if the given bundle exists
+// on a different master
+func (c *ClusterBundleHandler) Delete(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id := vars["id"]
+
+	masters, err := c.getMasterNodes()
+	if err != nil {
+		writeJSONError(w, http.StatusInternalServerError, fmt.Errorf("unable to get list of masters: %s", err))
+		return
+	}
+
+	ctx := context.Background()
+
+	found := false
+	for _, n := range masters {
+		err = c.client.Delete(ctx, n.baseURL, id)
+		if err != nil {
+			// some errors tell us the bundle was found on a master but something else was wrong so we end and return an error status
+			// but a NotFound error means it should keep going
+			switch err.(type) {
+			case *DiagnosticsBundleNotCompletedError:
+				w.WriteHeader(http.StatusNotModified)
+				return
+			case *DiagnosticsBundleUnreadableError:
+				writeJSONError(w, http.StatusInternalServerError, err)
+				return
+			}
+		}
+		if err == nil {
+			found = true
+		}
+	}
+
+	if !found {
+		writeJSONError(w, http.StatusNotFound, fmt.Errorf("bundle %s not found on any master", id))
+	}
+}
+
+// Download will download the given bundle, proxying the call to the appropriate master
+func (c *ClusterBundleHandler) Download(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id := vars["id"]
+
+	// TODO: for this one specifically, it would be ideal to detect if the bundle exists on the calling master
+	// first since then we can skip the intermediate download
+	masters, err := c.getMasterNodes()
+	if err != nil {
+		writeJSONError(w, http.StatusInternalServerError, fmt.Errorf("unable to get list of masters: %s", err))
+		return
+	}
+
+	ctx := context.Background()
+
+	var masterWithBundle node
+	found := false
+	for _, n := range masters {
+		_, err = c.client.Status(ctx, n.baseURL, id)
+		if err != nil {
+			switch err.(type) {
+			case *DiagnosticsBundleUnreadableError:
+				writeJSONError(w, http.StatusInternalServerError, err)
+				return
+			case *DiagnosticsBundleNotCompletedError:
+				writeJSONError(w, http.StatusNotFound, err)
+				return
+			case *DiagnosticsBundleNotFoundError:
+				continue
+			}
+		} else {
+			masterWithBundle = n
+			found = true
+			break
+		}
+	}
+	if !found {
+		writeJSONError(w, http.StatusNotFound, &DiagnosticsBundleNotFoundError{id: id})
+		return
+	}
+
+	bundleDir, err := ioutil.TempDir("", "bundle-")
+	if err != nil {
+		writeJSONError(w, http.StatusInternalServerError, fmt.Errorf("error opening temp file to download bundle %s", err))
+		return
+	}
+	defer os.RemoveAll(bundleDir)
+
+	bundleFilename := filepath.Join(bundleDir, "bundle.zip")
+
+	err = c.client.GetFile(ctx, masterWithBundle.baseURL, id, bundleFilename)
+	if err != nil {
+		writeJSONError(w, http.StatusInternalServerError, fmt.Errorf("error downloading bundle: %s", err))
+		return
+	}
+
+	data, err := ioutil.ReadFile(bundleFilename)
```

Comment:
This should probably be a `io.Copy(w, bundleFile)`.

Reply:
```http.ServeFile(w, r, bundleFilename)```

Reply:
https://github.com/dcos/dcos-diagnostics/pull/140